### PR TITLE
feat: support reading in quotaProjectId for billing

### PR DESCRIPTION
**Pull request by codyoss**

### CODE REVIEW

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 1]
```diff
@@ -166,6 +168,24 @@ public static GoogleCredentials fromStream(
             fileType, USER_FILE_TYPE, SERVICE_ACCOUNT_FILE_TYPE));
   }
 
+  /**
+   * Adds quota project ID to requestMetadata if present.
+   *
+   * @return a new map with quotaProjectId added if needed, or else returns the original
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: This is weird: sometimes it's a copy and sometimes it isn't? Pick one or the other. If this is only used internally, it's probably OK to modify the map.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 2]
```diff
@@ -75,11 +75,7 @@
 import java.security.SignatureException;
 import java.security.spec.InvalidKeySpecException;
 import java.security.spec.PKCS8EncodedKeySpec;
-import java.util.Collection;
-import java.util.Date;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
+import java.util.*;
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: no wild cards

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 3]
```diff
@@ -119,6 +116,8 @@
    * @param tokenServerUri URI of the end point that provides tokens.
    * @param serviceAccountUser Email of the user account to impersonate, if delegating domain-wide
    *     authority to the service account.
+   * @param projectId The project used for billing.
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: nit: The --> the
and no period when it isn't a complete sentence

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 4]
```diff
@@ -291,6 +295,49 @@ public static ServiceAccountCredentials fromPkcs8(
         null);
   }
 
+  /**
+   * Factory with minimum identifying information and custom transport using PKCS#8 for the private
+   * key.
+   *
+   * @param clientId Client ID of the service account from the console. May be null.
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: client ID
and similar below

File: oauth2_http/java/com/google/auth/oauth2/UserCredentials.java [hunk 5]
```diff
@@ -52,9 +52,7 @@
 import java.io.InputStream;
 import java.io.ObjectInputStream;
 import java.net.URI;
-import java.util.Date;
-import java.util.Map;
-import java.util.Objects;
+import java.util.*;
```
Comment (elharo) [inline oauth2_http/java/com/google/auth/oauth2/UserCredentials.java]: no wild cards

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java [hunk 6]
```diff
@@ -338,7 +339,12 @@ public void fromJSON_getProjectId() throws IOException {
     transportFactory.transport.addServiceAccount(SA_CLIENT_EMAIL, ACCESS_TOKEN);
     GenericJson json =
         writeServiceAccountJson(
-            SA_CLIENT_ID, SA_CLIENT_EMAIL, SA_PRIVATE_KEY_PKCS8, SA_PRIVATE_KEY_ID, PROJECT_ID);
+            SA_CLIENT_ID,
```
Comment (elharo) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java]: SA == SERVICE_ACCOUNT?
avoid abbreviations

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java (hunk 1 above)
Comment (codyoss) [reply oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: upstream the map unmodifiable. I will just always make a new map in this case then.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java (hunk 6 above)
Comment (codyoss) [reply oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java]: I would assume so, I did not add this, but I can expand.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 7]
```diff
@@ -291,6 +299,49 @@ public static ServiceAccountCredentials fromPkcs8(
...
+   * @param transportFactory HTTP transport factory, creates the transport used to get access
+   *     tokens.
+   * @param tokenServerUri URI of the end point that provides tokens.
+   * @param serviceAccountUser the email of the user account to impersonate, if delegating
+   *     domain-wide authority to the service account.
+   * @param quotaProjectId the project used for quota and billing purposes. May be null.
+   * @return New ServiceAccountCredentials created from a private key.
+   * @throws IOException if the credential cannot be created from the private key.
+   */
+  public static ServiceAccountCredentials fromPkcs8(
```
Comment (chingor13) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: I'm not a huge fan of adding more and more of these helpers for each combination of options. We should consider our options around this design for 1.0.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java [hunk 8]
```diff
@@ -1032,6 +1054,9 @@ static GenericJson writeServiceAccountJson(
     if (projectId != null) {
       json.put("project_id", projectId);
     }
+    if (quotaProjectId != null) {
```
Comment (chingor13) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java]: is it possible for the service account json file to have a null entry and is it worth writing a test case for it?

File: oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java [hunk 9]
```diff
@@ -166,6 +172,22 @@ public static GoogleCredentials fromStream(
...
   }
 
+  /**
+   * Adds quota project ID to requestMetadata if present.
+   *
+   * @return a new map with quotaProjectId added if needed
+   */
+  static Map<String, List<String>> addQuotaProjectIdToRequestMetadata(
+      String quotaProjectId, Map<String, List<String>> requestMetadata) {
+    Preconditions.checkNotNull(requestMetadata, "requestMetadata");
```
Comment (chingor13) [inline oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java]: nit: this exception message is not particularly useful. Let's make it better or remove the message.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 7 above)
Comment (codyoss) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: agreed. Maybe use a builder eventually.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java (hunk 8 above)
Comment (codyoss) [reply oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java]: It is, until this feature is live, this value will always be null. This is being implicitly tested via other tests that don't set this field. I am not sure I could add much more value with another test. Unless I am missing something?

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 7 above)
Comment (chingor13) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: IMO, if this is not necessary for our use case, let's not add it if we're considering deprecating it soon.

Comment (codyoss) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: fair, will revert these changes.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 10]
```diff
@@ -203,7 +209,32 @@ public static ServiceAccountJwtAccessCredentials fromPkcs8(
...
+   * @param clientId Client ID of the service account from the console. May be null.
+   * @param clientEmail Client email address of the service account from the console.
+   * @param privateKeyPkcs8 RSA private key object for the service account in PKCS#8 format.
+   * @param privateKeyId Private key identifier for the service account. May be null.
+   * @param defaultAudience Audience to use if not provided by transport. May be null.
+   * @param quotaProjectId The project used for quota and billing purposes. May be null.
+   * @return New ServiceAccountJwtAcceessCredentials created from a private key.
+   * @throws IOException if the credential cannot be created from the private key.
+   */
+  public static ServiceAccountJwtAccessCredentials fromPkcs8(
```
Comment (chingor13) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: This one is also unnecessary

Comment (chingor13) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: Is this one actually needed?

Comment (codyoss) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java]: I would say yes actually for this one. I moved it to package scope though as to not expose more surface to the api.

**Pull request merged**